### PR TITLE
LPS-183803 - Select file pop-up window from upload field does not have any context when editing a form entry

### PR DESCRIPTION
**Pull request by MirnaGama**
_**Steps to reproduce:**_
1. Log in as Admin
2. Create a new form with an upload field in it
3. Drag this form on any page
4. Fill the form with a dummy file and submit
5. Navigate to Site Menu > Content and Data > Forms > Options for the test form > View Entries > Options for the entry > Edit > Click on one of the Upload Field Select button
6. See the pop-up window without options

Expected Result: The user can choose and upload a new file successfully.

Actual Results: The pop-up window comes up but does not have nay options only the title is visible. 

_**Solution proposed:**_

I did some investigation and discovered that there is no request being made after we click the upload file selection button. This happened because the variable responsible for storing the request URL (itemSelectorURL) was null, so the request could not be made at all.

This variable is created in a back-end class, and depends on the field's context **not being read-only** and having a valid **groupId**. After debugging I found that the groupId was not being set correctly in the field context.

### CODE REVIEW

**Comment by MirnaGama** [comment]
ci:stop

**Comment by MirnaGama** [comment]
ci:test:sf

**Comment by MirnaGama** [comment]
ci:test:relevant

**Comment by MirnaGama** [comment]
ci:test:forms

**Comment by carolmariaabb** [comment]
@MirnaGama There's one test failing related to your changes: [com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest](https://test-1-1.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/165993/testReport/com.liferay.dynamic.data.mapping.form.field.type.internal.document.library/DocumentLibraryDDMFormFieldTemplateContextContributorTest). Could you please fix it? Thanks!

**Comment by MirnaGama** [comment]
ci:test:sf

**Comment by MirnaGama** [comment]
ci:test:relevant

**Comment by MirnaGama** [comment]
ci:test:forms

**Comment by MirnaGama** [comment]
Done @carolmariaabb ! I ran the test class in my environment and everything worked just fine :grin: 

```
com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest STANDARD_OUT
    Loading jar:file:/home/me/git/7.4.13-u64/.m2/com/liferay/portal/com.liferay.portal.impl/68.0.2-SNAPSHOT/com.liferay.portal.impl-68.0.2-SNAPSHOT.jar!/system.properties
    Loading jar:file:/home/me/bundles/7.4-u64/tomcat-9.0.71/webapps/ROOT/WEB-INF/shielded-container-lib/portal-impl.jar!/system.properties
    Loading jar:file:/home/me/git/7.4.13-u64/.m2/com/liferay/portal/com.liferay.portal.impl/68.0.2-SNAPSHOT/com.liferay.portal.impl-68.0.2-SNAPSHOT.jar!/portal.properties

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersShouldUseFileEntryTitle STARTED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersShouldUseFileEntryTitle PASSED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersForGuestUser STARTED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersForGuestUser PASSED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersWithNullGroupShouldContainItemSelectorURL STARTED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersWithNullGroupShouldContainItemSelectorURL PASSED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersShouldContainFileEntryURL STARTED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersShouldContainFileEntryURL PASSED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersForAllowedGuestUser STARTED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersForAllowedGuestUser PASSED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testDDMFormPortletItemSelector STARTED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testDDMFormPortletItemSelector PASSED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersForSignedInUser STARTED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersForSignedInUser PASSED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersForUserWithoutPermission STARTED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersForUserWithoutPermission PASSED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersShouldUseExistingItemSelectorURL STARTED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersShouldUseExistingItemSelectorURL PASSED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersShouldUseExistingGuestUploadURL STARTED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersShouldUseExistingGuestUploadURL PASSED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersShouldContainMaximumRepetitions STARTED

com.liferay.dynamic.data.mapping.form.field.type.internal.document.library.DocumentLibraryDDMFormFieldTemplateContextContributorTest > testGetParametersShouldContainMaximumRepetitions PASSED


BUILD SUCCESSFUL in 3m 53s
296 actionable tasks: 2 executed, 1 from cache, 293 up-to-date   
```

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/java/com/liferay/dynamic/data/mapping/form/renderer/internal/DDMFormFieldTemplateContextFactory.java [hunk 1]
```diff
@@ -111,6 +115,13 @@ public List<Object> create() {
 			_ddmFormFieldValues, StringPool.BLANK);
 	}
 
+	protected ThemeDisplay getThemeDisplay(
+		HttpServletRequest httpServletRequest) {
+
+		return (ThemeDisplay)httpServletRequest.getAttribute(
+			WebKeys.THEME_DISPLAY);
+	}
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/java/com/liferay/dynamic/data/mapping/form/renderer/internal/DDMFormFieldTemplateContextFactory.java:123]: @MirnaGama we actually don't need this method as it is only used once.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/java/com/liferay/dynamic/data/mapping/form/renderer/internal/DDMFormFieldTemplateContextFactory.java [hunk 2]
```diff
@@ -161,6 +172,18 @@ private DDMFormFieldRenderingContext _createDDDMFormFieldRenderingContext(
...
+		long groupId = GetterUtil.getLong(
+			ddmFormFieldRenderingContext.getProperty("groupId"));
+
+		if (groupId == 0) {
+			ThemeDisplay themeDisplay = getThemeDisplay(
+				ddmFormFieldRenderingContext.getHttpServletRequest());
+
+			ddmFormFieldRenderingContext.setProperty(
+				"groupId", themeDisplay.getScopeGroupId());
+		}
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/java/com/liferay/dynamic/data/mapping/form/renderer/internal/DDMFormFieldTemplateContextFactory.java:185]: @MirnaGama please change to:

```
long groupId = _ddmFormRenderingContext.getGroupId();

if (groupId == 0) {
    HttpServletRequest httpServletRequest = ddmFormFieldRenderingContext.getHttpServletRequest();
    
    ThemeDisplay themeDisplay = (ThemeDisplay)httpServletRequest.getAttribute(WebKeys.THEME_DISPLAY);
    
    groupId = themeDisplay.getScopeGroupId();
}

ddmFormFieldRenderingContext.setProperty("groupId", groupId);
```

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/display/context/DDMFormAdminDisplayContext.java [hunk 3]
```diff
@@ -714,7 +714,7 @@ public JSONObject getFormBuilderContextJSONObject() throws PortalException {
 		DDMFormBuilderContextRequest ddmFormBuilderContextRequest =
 			DDMFormBuilderContextRequest.with(
 				null, themeDisplay.getRequest(), themeDisplay.getResponse(),
-				LocaleUtil.fromLanguageId(getDefaultLanguageId()), true);
+				LocaleUtil.fromLanguageId(getDefaultLanguageId()), false);
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/display/context/DDMFormAdminDisplayContext.java:717]: @MirnaGama why did you have to change this code? Aren't we changing the expected behavior when the context **needs** to be read-only? 

For example, I noticed that the `getFormBuilderContextJSONObject` method is called when editing the form instance in the form builder page, so in this case the context **must** be read-only. 

image

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

**Comment by MirnaGama** [comment]
@carolmariaabb Done !

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/java/com/liferay/dynamic/data/mapping/form/renderer/internal/DDMFormFieldTemplateContextFactory.java (hunk 2 above)
Comment (MirnaGama) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/java/com/liferay/dynamic/data/mapping/form/renderer/internal/DDMFormFieldTemplateContextFactory.java]: done !

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/display/context/DDMFormAdminDisplayContext.java (hunk 3 above)
Comment (MirnaGama) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/display/context/DDMFormAdminDisplayContext.java:717]: @carolmariaabb 

In the form entries area, the form context is always read-only and this prevents the itemSelector URL from being filled and returned, as we can see [in this code](https://github.com/liferay/liferay-portal/blob/master/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/java/com/liferay/dynamic/data/mapping/form/field/type/internal/document/library/DocumentLibraryDDMFormFieldTemplateContextContributor.java#L602) from _getUploadParameters method. In this case, could we remove this verification instead ?

Comment (carolmariaabb) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/display/context/DDMFormAdminDisplayContext.java:717]: @MirnaGama I've tested your code and it seems like you don't need this change to fix the issue. The commit related to the `groupId` change it's enough. Could you please double check?

[FixedIssue.webm](https://github.com/liferay-forms/liferay-portal/assets/34186632/205b2a12-38d1-4742-b8d6-60b377b18495)

Comment (MirnaGama) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/display/context/DDMFormAdminDisplayContext.java]: You're right @carolmariaabb 
Thanks, I've resent a new pull request with only the necessary commits: https://github.com/liferay-forms/liferay-portal/pull/2212

**Pull request closed without merging**